File: Android/app/src/main/java/com/swarnkary/donezy/CreateTrackerSheet.kt
package com.swarnkary.donezy

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CalendarMonth
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Schedule
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableLongStateOf
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import java.time.Instant
import java.time.LocalDate
import java.time.LocalTime
import java.time.ZoneId
import java.time.ZonedDateTime
import java.util.Calendar

/**
 * Result of the create/edit sheet's reminder section.
 *  - reminderAt = null + recurrence = None  → no reminder
 *  - reminderAt non-null + recurrence = None → one-shot at that time
 *  - both set                               → first fire + repeat
 */
data class ReminderConfig(val reminderAt: Long?, val recurrence: Recurrence)

data class TrackerSaveResult(
    val name: String,
    val category: String,
    val notes: String,
    val reminder: ReminderConfig,
    val weeklyGoal: Int
)

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CreateTrackerSheet(
    sheetState: SheetState,
    onDismiss: () -> Unit,
    onSave: (TrackerSaveResult) -> Unit,
    initial: Hobby? = null,           // non-null = edit mode
    prefill: HobbyTemplate? = null    // populates fields when no `initial`
) {
    val isEdit = initial != null

    var name     by remember { mutableStateOf(initial?.name ?: prefill?.trackerName ?: "") }
    var category by remember { mutableStateOf(categoryFor(initial?.category ?: prefill?.category ?: "General")) }
    var notes    by remember { mutableStateOf(initial?.notes ?: prefill?.notes ?: "") }
    var weeklyGoal by remember { mutableIntStateOf(initial?.weeklyGoal ?: 0) }

    // Reminder state — preloaded from existing tracker or template default
    var reminderAt by remember {
        mutableStateOf(
            initial?.nextReminderAt
                ?: prefill?.reminderHours?.takeIf { it > 0L }
                    ?.let { System.currentTimeMillis() + it * HOUR_MS }
        )
    }
    var recurrence by remember {
        mutableStateOf(
            initial?.recurrence
                ?: prefill?.reminderHours?.takeIf { it > 0L }?.let { Recurrence.Hourly(it) }
                ?: Recurrence.None
        )
    }

    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        dragHandle = { BottomSheetDefaults.DragHandle() },
        containerColor = MaterialTheme.colorScheme.surface
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(horizontal = 20.dp)
                .padding(bottom = 40.dp),
            verticalArrangement = Arrangement.spacedBy(16.dp)
        ) {
            // Header
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically
            ) {
                Text(
                    if (isEdit) "Edit tracker" else "New tracker",
                    style = MaterialTheme.typography.titleLarge,
                    fontWeight = FontWeight.Bold
                )
                IconButton(onClick = onDismiss) {
                    Icon(Icons.Default.Close, contentDescription = "Close")
                }
            }

            OutlinedTextField(
                value = name,
                onValueChange = { name = it },
                label = { Text("Tracker name") },
                singleLine = true,
                modifier = Modifier.fillMaxWidth(),
                shape = MaterialTheme.shapes.medium
            )

            CategoryDropdown(selected = category, onSelected = { category = it })

            AnimatedVisibility(visible = category.description.isNotBlank(), enter = fadeIn(), exit = fadeOut()) {
                Text(
                    category.description,
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            }

            OutlinedTextField(
                value = notes,
                onValueChange = { notes = it },
                label = { Text("Plan, baseline, or care notes") },
                placeholder = { Text(category.starterNotes, style = MaterialTheme.typography.bodySmall) },
                modifier = Modifier.fillMaxWidth(),
                minLines = 3,
                shape = MaterialTheme.shapes.medium
            )

            ReminderSection(
                reminderAt = reminderAt,
                recurrence = recurrence,
                onChange = { ts, rule -> reminderAt = ts; recurrence = rule }
            )

            WeeklyGoalSection(
                value = weeklyGoal,
                onChange = { weeklyGoal = it }
            )

            Spacer(Modifier.height(4.dp))

            Button(
                onClick = {
                    onSave(
                        TrackerSaveResult(
                            name = name,
                            category = category.label,
                            notes = notes,
                            reminder = ReminderConfig(reminderAt, recurrence),
                            weeklyGoal = weeklyGoal
                        )
                    )
                },
                enabled = name.isNotBlank(),
                modifier = Modifier.fillMaxWidth().height(52.dp),
                shape = MaterialTheme.shapes.large
            ) {
                Text(
                    if (isEdit) "Save changes" else "Add tracker",
                    style = MaterialTheme.typography.titleSmall,
                    fontWeight = FontWeight.SemiBold
                )
            }
        }
    }
}

// ─── Weekly goal ──────────────────────────────────────────────────────────────

@Composable
private fun WeeklyGoalSection(value: Int, onChange: (Int) -> Unit) {
    Text("Weekly goal", style = MaterialTheme.typography.titleSmall, fontWeight = FontWeight.SemiBold)
    Text(
        "Earn the 🎯 badge each week you hit this many logs.",
        style = MaterialTheme.typography.bodySmall,
        color = MaterialTheme.colorScheme.onSurfaceVariant
    )
    val options = listOf(0, 1, 2, 3, 5, 7, 10)
    androidx.compose.foundation.lazy.LazyRow(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        items(options) { n ->
            FilterChip(
                selected = value == n,
                onClick = { onChange(n) },
                label = { Text(if (n == 0) "Off" else "$n / wk") }
            )
        }
    }
}

// ─── Reminder Section ─────────────────────────────────────────────────────────

@OptIn(ExperimentalMaterial3Api::class)
@Composable
private fun ReminderSection(
    reminderAt: Long?,
    recurrence: Recurrence,
    onChange: (Long?, Recurrence) -> Unit
) {
    var showDatePicker by remember { mutableStateOf(false) }
    var showTimePicker by remember { mutableStateOf(false) }
    val zone = ZoneId.systemDefault()

    Text("Reminder", style = MaterialTheme.typography.titleSmall, fontWeight = FontWeight.SemiBold)

    // Quick relative chips — "in 1h, 3h, 1d, 1w, …"
    LazyRow(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        items(reminderOptions) { option ->
            val selectedHours = if (reminderAt != null && recurrence is Recurrence.Hourly)
                recurrence.hours else null
            val matchesNone = option.hours == 0L && reminderAt == null
            val matches = matchesNone || option.hours == selectedHours
            FilterChip(
                selected = matches,
                onClick = {
                    if (option.hours == 0L) {
                        onChange(null, Recurrence.None)
                    } else {
                        onChange(System.currentTimeMillis() + option.hours * HOUR_MS, Recurrence.None)
                    }
                },
                label = {
                    Column(horizontalAlignment = Alignment.CenterHorizontally) {
                        Text(option.label, fontWeight = if (matches) FontWeight.SemiBold else FontWeight.Normal)
                        Text(option.description, style = MaterialTheme.typography.labelSmall)
                    }
                }
            )
        }
    }

    // Absolute datetime row
    Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        OutlinedButton(
            onClick = { showDatePicker = true },
            modifier = Modifier.weight(1f),
            shape = MaterialTheme.shapes.large
        ) {
            Icon(Icons.Default.CalendarMonth, contentDescription = null, modifier = Modifier.size(16.dp))
            Spacer(Modifier.width(6.dp))
            Text(reminderAt?.let { formatDateShort(it) } ?: "Pick date")
        }
        OutlinedButton(
            onClick = { if (reminderAt != null) showTimePicker = true else showDatePicker = true },
            modifier = Modifier.weight(1f),
            shape = MaterialTheme.shapes.large
        ) {
            Icon(Icons.Default.Schedule, contentDescription = null, modifier = Modifier.size(16.dp))
            Spacer(Modifier.width(6.dp))
            Text(reminderAt?.let {
                val t = ZonedDateTime.ofInstant(Instant.ofEpochMilli(it), zone).toLocalTime()
                "%02d:%02d".format(t.hour, t.minute)
            } ?: "Pick time")
        }
    }

    // Recurrence choice
    AnimatedVisibility(visible = reminderAt != null, enter = fadeIn(), exit = fadeOut()) {
        Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
            Text("Repeat", style = MaterialTheme.typography.labelLarge,
                color = MaterialTheme.colorScheme.onSurfaceVariant)
            RecurrencePicker(
                current = recurrence,
                anchor = reminderAt ?: System.currentTimeMillis(),
                onChange = { onChange(reminderAt, it) }
            )
        }
    }

    if (showDatePicker) {
        val initialMillis = reminderAt ?: (System.currentTimeMillis() + DAY_MS)
        val state = rememberDatePickerState(initialSelectedDateMillis = initialMillis)
        DatePickerDialog(
            onDismissRequest = { showDatePicker = false },
            confirmButton = {
                TextButton(onClick = {
                    val ms = state.selectedDateMillis ?: initialMillis
                    val zdt = ZonedDateTime.ofInstant(Instant.ofEpochMilli(ms), ZoneId.of("UTC"))
                        .toLocalDate()
                    val time = reminderAt
                        ?.let { ZonedDateTime.ofInstant(Instant.ofEpochMilli(it), zone).toLocalTime() }
                        ?: LocalTime.of(9, 0)
                    val combined = ZonedDateTime.of(zdt, time, zone).toInstant().toEpochMilli()
                    onChange(combined, recurrence)
                    showDatePicker = false
                }) { Text("OK") }
            },
            dismissButton = {
                TextButton(onClick = { showDatePicker = false }) { Text("Cancel") }
            }
        ) {
            DatePicker(state = state)
        }
    }

    if (showTimePicker && reminderAt != null) {
        val current = ZonedDateTime.ofInstant(Instant.ofEpochMilli(reminderAt), zone).toLocalTime()
        val state = rememberTimePickerState(initialHour = current.hour, initialMinute = current.minute, is24Hour = false)
        AlertDialog(
            onDismissRequest = { showTimePicker = false },
            confirmButton = {
                TextButton(onClick = {
                    val date = ZonedDateTime.ofInstant(Instant.ofEpochMilli(reminderAt), zone).toLocalDate()
                    val combined = ZonedDateTime.of(date, LocalTime.of(state.hour, state.minute), zone)
                        .toInstant().toEpochMilli()
                    onChange(combined, recurrence)
                    showTimePicker = false
                }) { Text("OK") }
            },
            dismissButton = {
                TextButton(onClick = { showTimePicker = false }) { Text("Cancel") }
            },
            title = { Text("Pick time") },
            text = { TimePicker(state = state) }
        )
    }
}

// ─── Recurrence picker ────────────────────────────────────────────────────────

@Composable
private fun RecurrencePicker(
    current: Recurrence,
    anchor: Long,
    onChange: (Recurrence) -> Unit
) {
    val zone = ZoneId.systemDefault()
    val anchorDate = ZonedDateTime.ofInstant(Instant.ofEpochMilli(anchor), zone).toLocalDate()

    val choices = listOf<Pair<String, Recurrence>>(
        "None"    to Recurrence.None,
        "Hourly"  to (current as? Recurrence.Hourly ?: Recurrence.Hourly(24L)),
        "Daily"   to Recurrence.Daily,
        "Weekly"  to (current as? Recurrence.Weekly ?: Recurrence.Weekly(1 shl ((anchorDate.dayOfWeek.value - 1)))),
        "Monthly" to (current as? Recurrence.Monthly ?: Recurrence.Monthly(anchorDate.dayOfMonth))
    )

    LazyRow(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        items(choices) { (label, rule) ->
            val selected = current.javaClass == rule.javaClass &&
                (current !is Recurrence.None || rule is Recurrence.None)
            FilterChip(
                selected = selected,
                onClick = { onChange(rule) },
                label = { Text(label) }
            )
        }
    }

    when (current) {
        is Recurrence.Hourly -> HourlyEditor(current.hours) { onChange(Recurrence.Hourly(it)) }
        is Recurrence.Weekly -> WeeklyEditor(current.dayMask) { onChange(Recurrence.Weekly(it)) }
        is Recurrence.Monthly -> Text(
            "Day ${current.dayOfMonth} of each month",
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant
        )
        else -> { /* no extra controls */ }
    }
}

@Composable
private fun HourlyEditor(hours: Long, onChange: (Long) -> Unit) {
    val presets = listOf(1L, 3L, 6L, 12L, 24L, 48L, 72L, 168L, 336L, 720L)
    LazyRow(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
        items(presets) { p ->
            FilterChip(
                selected = hours == p,
                onClick = { onChange(p) },
                label = { Text(reminderOptions.firstOrNull { it.hours == p }?.label ?: "${p}h") }
            )
        }
    }
}

@Composable
private fun WeeklyEditor(dayMask: Int, onChange: (Int) -> Unit) {
    val labels = listOf("M", "T", "W", "T", "F", "S", "S")
    Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
        labels.forEachIndexed { i, label ->
            val selected = dayMask and (1 shl i) != 0
            Box(
                modifier = Modifier
                    .size(36.dp)
                    .clip(RoundedCornerShape(999.dp))
                    .background(
                        if (selected) MaterialTheme.colorScheme.primary
                        else MaterialTheme.colorScheme.surfaceVariant
                    )
                    .clickable {
                        val newMask = dayMask xor (1 shl i)
                        onChange(newMask.coerceAtLeast(0))
                    },
                contentAlignment = Alignment.Center
            ) {
                Text(
                    label,
                    color = if (selected) MaterialTheme.colorScheme.onPrimary
                        else MaterialTheme.colorScheme.onSurface,
                    fontWeight = FontWeight.SemiBold
                )
            }
        }
    }
}

// ─── Category Dropdown ────────────────────────────────────────────────────────

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CategoryDropdown(selected: CategoryOption, onSelected: (CategoryOption) -> Unit) {
    var expanded by remember { mutableStateOf(false) }

    ExposedDropdownMenuBox(expanded = expanded, onExpandedChange = { expanded = !expanded }) {
        OutlinedTextField(
            value = "${selected.emoji}  ${selected.label}",
            onValueChange = {},
            readOnly = true,
            label = { Text("Category") },
            trailingIcon = { ExposedDropdownMenuDefaults.TrailingIcon(expanded = expanded) },
            modifier = Modifier
                .menuAnchor(MenuAnchorType.PrimaryNotEditable)
                .fillMaxWidth(),
            shape = MaterialTheme.shapes.medium
        )
        ExposedDropdownMenu(expanded = expanded, onDismissRequest = { expanded = false }) {
            hobbyCategories.forEach { option ->
                DropdownMenuItem(
                    text = {
                        Row(
                            horizontalArrangement = Arrangement.spacedBy(10.dp),
                            verticalAlignment = Alignment.CenterVertically
                        ) {
                            Text(option.emoji)
                            Column {
                                Text(option.label, fontWeight = FontWeight.SemiBold)
                                Text(
                                    option.description,
                                    style = MaterialTheme.typography.bodySmall,
                                    color = MaterialTheme.colorScheme.onSurfaceVariant
                                )
                            }
                        }
                    },
                    onClick = { onSelected(option); expanded = false }
                )
            }
        }
    }
}

// ─── Template Deck ────────────────────────────────────────────────────────────

@Composable
fun TemplateDeck(onApply: (HobbyTemplate) -> Unit) {
    Column(verticalArrangement = Arrangement.spacedBy(10.dp)) {
        SectionHeader("Starter templates", subtitle = "Tap to prefill")
        LazyRow(
            horizontalArrangement = Arrangement.spacedBy(12.dp),
            contentPadding = PaddingValues(end = 4.dp)
        ) {
            items(hobbyTemplates) { template ->
                val cat = categoryFor(template.category)
                ElevatedCard(
                    onClick = { onApply(template) },
                    modifier = Modifier.width(220.dp),
                    shape = MaterialTheme.shapes.large,
                    colors = CardDefaults.elevatedCardColors(containerColor = MaterialTheme.colorScheme.surface),
                    elevation = CardDefaults.elevatedCardElevation(defaultElevation = 2.dp)
                ) {
                    Column(Modifier.padding(14.dp), verticalArrangement = Arrangement.spacedBy(8.dp)) {
                        CategoryBadge(cat)
                        Text(template.title, style = MaterialTheme.typography.titleSmall, fontWeight = FontWeight.SemiBold)
                        Text(
                            template.notes,
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                            maxLines = 3
                        )
                        Text(
                            "Default: ${reminderLabel(template.reminderHours)}",
                            style = MaterialTheme.typography.labelSmall,
                            color = MaterialTheme.colorScheme.primary
                        )
                    }
                }
            }
        }
    }
}

File: MacOs/Donezy/CreateTrackerSheet.swift
import SwiftUI

/// Result of the create/edit sheet's reminder section.
struct ReminderConfig: Equatable {
    let reminderAt: Int64?
    let recurrence: Recurrence
}

struct TrackerSaveResult: Equatable {
    let name: String
    let category: String
    let notes: String
    let reminder: ReminderConfig
    let weeklyGoal: Int
}

/// Bottom-sheet form for creating or editing a tracker. Port of `CreateTrackerSheet`.
struct CreateTrackerSheet: View {
    var initial: Hobby?               // non-nil = edit mode
    var prefill: HobbyTemplate?       // populates fields when no `initial`
    let onSave: (TrackerSaveResult) -> Void
    let onDismiss: () -> Void

    @Environment(\.theme) private var theme

    @State private var name: String
    @State private var category: CategoryOption
    @State private var notes: String
    @State private var weeklyGoal: Int
    @State private var reminderAt: Int64?
    @State private var recurrence: Recurrence

    init(initial: Hobby? = nil, prefill: HobbyTemplate? = nil,
         onSave: @escaping (TrackerSaveResult) -> Void, onDismiss: @escaping () -> Void) {
        self.initial = initial
        self.prefill = prefill
        self.onSave = onSave
        self.onDismiss = onDismiss

        _name = State(initialValue: initial?.name ?? prefill?.trackerName ?? "")
        _category = State(initialValue: categoryFor(initial?.category ?? prefill?.category ?? "General"))
        _notes = State(initialValue: initial?.notes ?? prefill?.notes ?? "")
        _weeklyGoal = State(initialValue: initial?.weeklyGoal ?? 0)

        let initialReminder: Int64? = {
            if let r = initial?.nextReminderAt { return r }
            if let h = prefill?.reminderHours, h > 0 { return nowMillis() + h * HOUR_MS }
            return nil
        }()
        _reminderAt = State(initialValue: initialReminder)

        let initialRecurrence: Recurrence = {
            if let r = initial?.recurrence { return r }
            if let h = prefill?.reminderHours, h > 0 { return .hourly(hours: h) }
            return .none
        }()
        _recurrence = State(initialValue: initialRecurrence)
    }

    private var isEdit: Bool { initial != nil }

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(alignment: .leading, spacing: 16) {
                    TextField("Tracker name", text: $name)
                        .textFieldStyle(.roundedBorder)
                        .autocorrectionDisabled()

                    CategoryDropdown(selected: $category)

                    if !category.description.isEmpty {
                        Text(category.description)
                            .font(.system(size: 12))
                            .foregroundColor(theme.onSurfaceVariant)
                    }

                    VStack(alignment: .leading, spacing: 4) {
                        Text("Plan, baseline, or care notes")
                            .font(.system(size: 12)).foregroundColor(theme.onSurfaceVariant)
                        TextEditor(text: $notes)
                            .frame(minHeight: 90)
                            .padding(6)
                            .autocorrectionDisabled()
                            .overlay(RoundedRectangle(cornerRadius: 8).strokeBorder(theme.onSurfaceVariant.opacity(0.4)))
                            .overlay(alignment: .topLeading) {
                                if notes.isEmpty {
                                    Text(category.starterNotes)
                                        .font(.system(size: 12))
                                        .foregroundColor(theme.onSurfaceVariant.opacity(0.6))
                                        .padding(.horizontal, 11).padding(.vertical, 14)
                                        .allowsHitTesting(false)
                                }
                            }
                    }

                    ReminderSection(reminderAt: $reminderAt, recurrence: $recurrence)

                    WeeklyGoalSection(value: $weeklyGoal)

                    Button {
                        onSave(TrackerSaveResult(
                            name: name, category: category.label, notes: notes,
                            reminder: ReminderConfig(reminderAt: reminderAt, recurrence: recurrence),
                            weeklyGoal: weeklyGoal))
                    } label: {
                        Text(isEdit ? "Save changes" : "Add tracker")
                            .font(.system(size: 14, weight: .semibold))
                            .frame(maxWidth: .infinity).frame(height: 52)
                    }
                    .buttonStyle(.borderedProminent)
                    .disabled(name.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)
                }
                .padding(20)
            }
            .navigationTitle(isEdit ? "Edit tracker" : "New tracker")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarTrailing) {
                    Button { onDismiss() } label: { Image(systemName: "xmark") }
                }
            }
        }
    }
}

// ─── Weekly goal ──────────────────────────────────────────────────────────────

private struct WeeklyGoalSection: View {
    @Binding var value: Int
    @Environment(\.theme) private var theme
    private let options = [0, 1, 2, 3, 5, 7, 10]
    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Text("Weekly goal").font(.system(size: 14, weight: .semibold))
            Text("Earn the 🎯 badge each week you hit this many logs.")
                .font(.system(size: 12)).foregroundColor(theme.onSurfaceVariant)
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 8) {
                    ForEach(options, id: \.self) { n in
                        FilterChip(text: n == 0 ? "Off" : "\(n) / wk", selected: value == n) { value = n }
                    }
                }
            }
        }
    }
}

// ─── Reminder Section ─────────────────────────────────────────────────────────

private struct ReminderSection: View {
    @Binding var reminderAt: Int64?
    @Binding var recurrence: Recurrence
    @Environment(\.theme) private var theme

    @State private var showDatePicker = false
    @State private var showTimePicker = false

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text("Reminder").font(.system(size: 14, weight: .semibold))

            // Quick relative chips
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 8) {
                    ForEach(reminderOptions) { option in
                        let selectedHours: Int64? = {
                            if reminderAt != nil, case .hourly(let h) = recurrence { return h }
                            return nil
                        }()
                        let matchesNone = option.hours == 0 && reminderAt == nil
                        let matches = matchesNone || option.hours == selectedHours
                        DualFilterChip(label: option.label, detail: option.description, selected: matches) {
                            if option.hours == 0 {
                                reminderAt = nil; recurrence = .none
                            } else {
                                reminderAt = nowMillis() + option.hours * HOUR_MS; recurrence = .none
                            }
                        }
                    }
                }
            }

            // Absolute date / time
            HStack(spacing: 8) {
                Button {
                    showDatePicker = true
                } label: {
                    Label(reminderAt.map { formatDateShort($0) } ?? "Pick date", systemImage: "calendar")
                        .frame(maxWidth: .infinity)
                }
                .buttonStyle(.bordered)

                Button {
                    if reminderAt != nil { showTimePicker = true } else { showDatePicker = true }
                } label: {
                    Label(timeLabel, systemImage: "clock")
                        .frame(maxWidth: .infinity)
                }
                .buttonStyle(.bordered)
            }

            if reminderAt != nil {
                VStack(alignment: .leading, spacing: 8) {
                    Text("Repeat").font(.system(size: 14, weight: .medium))
                        .foregroundColor(theme.onSurfaceVariant)
                    RecurrencePicker(current: $recurrence, anchor: reminderAt ?? nowMillis())
                }
            }
        }
        .sheet(isPresented: $showDatePicker) {
            DateTimePickerSheet(reminderAt: $reminderAt, mode: .date) { showDatePicker = false }
        }
        .sheet(isPresented: $showTimePicker) {
            DateTimePickerSheet(reminderAt: $reminderAt, mode: .time) { showTimePicker = false }
        }
    }

    private var timeLabel: String {
        guard let reminderAt else { return "Pick time" }
        let date = Date(timeIntervalSince1970: Double(reminderAt) / 1000.0)
        let comps = Calendar.current.dateComponents([.hour, .minute], from: date)
        return String(format: "%02d:%02d", comps.hour ?? 0, comps.minute ?? 0)
    }
}

/// Date or time picker presented as a sheet; writes back the merged instant.
///
/// Mirrors the Android `ReminderSection` dialogs:
///  • date mode  → a full month calendar (Material3 `DatePicker` equivalent). When no
///    reminder is set yet, the time defaults to 9:00 AM, exactly like Android.
///  • time mode  → an always-visible wheel clock, the closest iOS-native match to
///    Android's `TimePicker` dial; the existing date is preserved.
private struct DateTimePickerSheet: View {
    @Binding var reminderAt: Int64?
    enum Mode { case date, time }
    let mode: Mode
    let onClose: () -> Void

    @Environment(\.theme) private var theme
    @State private var selection: Date

    init(reminderAt: Binding<Int64?>, mode: Mode, onClose: @escaping () -> Void) {
        _reminderAt = reminderAt
        self.mode = mode
        self.onClose = onClose

        let calendar = Calendar.current
        if let existing = reminderAt.wrappedValue {
            // Editing an existing reminder — start from it.
            _selection = State(initialValue: Date(timeIntervalSince1970: Double(existing) / 1000.0))
        } else {
            // No reminder yet: default to tomorrow at 9:00 AM (matches Android).
            let tomorrow = calendar.date(byAdding: .day, value: 1, to: Date()) ?? Date()
            let nineAM = calendar.date(bySettingHour: 9, minute: 0, second: 0, of: tomorrow) ?? tomorrow
            _selection = State(initialValue: nineAM)
        }
    }

    var body: some View {
        NavigationStack {
            VStack {
                if mode == .date {
                    DatePicker("", selection: $selection, displayedComponents: .date)
                        .datePickerStyle(.graphical)
                        .labelsHidden()
                        .padding(.horizontal)
                    Spacer()
                } else {
                    Spacer()
                    DatePicker("", selection: $selection, displayedComponents: .hourAndMinute)
                        .datePickerStyle(.wheel)
                        .labelsHidden()
                    Spacer()
                }
            }
            .padding(.vertical)
            .navigationTitle(mode == .date ? "Pick date" : "Pick time")
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .cancellationAction) { Button("Cancel", action: onClose) }
                ToolbarItem(placement: .confirmationAction) {
                    Button("OK") {
                        reminderAt = Int64(selection.timeIntervalSince1970 * 1000.0)
                        onClose()
                    }
                }
            }
        }
        .presentationDetents(mode == .date ? [.large] : [.medium])
    }
}

// ─── Recurrence picker ────────────────────────────────────────────────────────

private struct RecurrencePicker: View {
    @Binding var current: Recurrence
    let anchor: Int64
    @Environment(\.theme) private var theme

    var body: some View {
        let anchorDate = Date(timeIntervalSince1970: Double(anchor) / 1000.0)
        let cal = Calendar.current
        // Mon=0 … Sun=6
        let anchorWeekdayBit = ((cal.component(.weekday, from: anchorDate) + 5) % 7)
        let anchorDay = cal.component(.day, from: anchorDate)

        let choices: [(String, Recurrence)] = [
            ("None", .none),
            ("Hourly", { if case .hourly = current { return current } else { return .hourly(hours: 24) } }()),
            ("Daily", .daily),
            ("Weekly", { if case .weekly = current { return current } else { return .weekly(dayMask: 1 << anchorWeekdayBit) } }()),
            ("Monthly", { if case .monthly = current { return current } else { return .monthly(dayOfMonth: anchorDay) } }()),
        ]

        VStack(alignment: .leading, spacing: 8) {
            ScrollView(.horizontal, showsIndicators: false) {
                HStack(spacing: 8) {
                    ForEach(choices, id: \.0) { label, rule in
                        FilterChip(text: label, selected: sameKind(current, rule)) { current = rule }
                    }
                }
            }

            switch current {
            case .hourly(let hours):
                HourlyEditor(hours: hours) { current = .hourly(hours: $0) }
            case .weekly(let mask):
                WeeklyEditor(dayMask: mask) { current = .weekly(dayMask: $0) }
            case .monthly(let day):
                Text("Day \(day) of each month")
                    .font(.system(size: 12)).foregroundColor(theme.onSurfaceVariant)
            default:
                EmptyView()
            }
        }
    }

    private func sameKind(_ a: Recurrence, _ b: Recurrence) -> Bool {
        switch (a, b) {
        case (.none, .none), (.hourly, .hourly), (.daily, .daily),
             (.weekly, .weekly), (.monthly, .monthly):
            return true
        default:
            return false
        }
    }
}

private struct HourlyEditor: View {
    let hours: Int64
    let onChange: (Int64) -> Void
    private let presets: [Int64] = [1, 3, 6, 12, 24, 48, 72, 168, 336, 720]
    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 6) {
                ForEach(presets, id: \.self) { p in
                    FilterChip(text: reminderOptions.first(where: { $0.hours == p })?.label ?? "\(p)h",
                               selected: hours == p) { onChange(p) }
                }
            }
        }
    }
}

private struct WeeklyEditor: View {
    let dayMask: Int
    let onChange: (Int) -> Void
    @Environment(\.theme) private var theme
    private let labels = ["M", "T", "W", "T", "F", "S", "S"]
    var body: some View {
        HStack(spacing: 6) {
            ForEach(0..<7, id: \.self) { i in
                let selected = dayMask & (1 << i) != 0
                Text(labels[i])
                    .font(.system(size: 14, weight: .semibold))
                    .foregroundColor(selected ? theme.onPrimary : theme.onSurface)
                    .frame(width: 36, height: 36)
                    .background(selected ? theme.primary : theme.surfaceVariant, in: Circle())
                    .onTapGesture { onChange(max(0, dayMask ^ (1 << i))) }
            }
        }
    }
}

// ─── Category Dropdown ────────────────────────────────────────────────────────

struct CategoryDropdown: View {
    @Binding var selected: CategoryOption
    @Environment(\.theme) private var theme
    var body: some View {
        Menu {
            ForEach(hobbyCategories, id: \.label) { option in
                Button {
                    selected = option
                    UIApplication.shared.sendAction(#selector(UIResponder.resignFirstResponder), to: nil, from: nil, for: nil)
                } label: {
                    if selected.label == option.label {
                        Label("\(option.emoji)  \(option.label)", systemImage: "checkmark")
                    } else {
                        Text("\(option.emoji)  \(option.label)")
                    }
                }
            }
        } label: {
            HStack {
                Text("\(selected.emoji)  \(selected.label)").foregroundColor(theme.onSurface)
                Spacer()
                Image(systemName: "chevron.up.chevron.down").font(.system(size: 12)).foregroundColor(theme.onSurfaceVariant)
            }
            .padding(12)
            .background(Color.clear)
            .overlay(RoundedRectangle(cornerRadius: 8).strokeBorder(theme.onSurfaceVariant.opacity(0.4)))
            .contentShape(Rectangle())
        }
        .simultaneousGesture(TapGesture().onEnded {
            UIApplication.shared.sendAction(#selector(UIResponder.resignFirstResponder), to: nil, from: nil, for: nil)
        })
    }
}
